File: XiePinyin/Logic/ChangeSet.cs
﻿using System;
using System.Collections.Generic;
using System.Diagnostics;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace XiePinyin.Logic
{
    [DebuggerDisplay("{DebugStr}")]
    class ChangeSet
    {
        [JsonProperty("lengthBefore")]
        public int LengthBefore;

        [JsonProperty("lengthAfter")]
        public int LengthAfter;

        /// <summary>
        /// Boxed integers for kept indexes, or <see cref="XieChar"/>s for insertions.
        /// </summary>
        [JsonProperty("items")]
        public readonly List<object> Items = new List<object>();

        [JsonIgnore]
        public string DebugStr
        {
            get { return ToDiagStr(); }
        }

        public static ChangeSet CreateIdent(int length)
        {
            var res = new ChangeSet { LengthBefore = length, LengthAfter = length };
            res.Items.Capacity = length;
            for (int i = 0; i < length; ++i) res.Items.Add(i);
            return res;
        }

        public static ChangeSet FromJson(string json)
        {
            var res = new ChangeSet();
            var obj = JObject.Parse(json);
            res.LengthBefore = (int)obj.GetValue("lengthBefore");
            res.LengthAfter = (int)obj.GetValue("lengthAfter");
            foreach (var itm in obj.GetValue("items"))
            {
                if (itm.Type == JTokenType.Integer) res.Items.Add((int)itm);
                else
                {
                    string hanzi = (string)(itm as JObject).GetValue("hanzi");
                    string pinyin = (string)(itm as JObject).GetValue("pinyin");
                    if (pinyin != null) res.Items.Add(new XieChar(hanzi, pinyin));
                    else res.Items.Add(new XieChar(hanzi));
                }    
            }
            return res;
        }

        public string SerializeJson()
        {
            return JsonConvert.SerializeObject(this);
        }

        public static ChangeSet FromDiagStr(string str)
        {
            var markIx = str.IndexOf('>');
            var lengthBeforeStr = str.Substring(0, markIx);
            int lengthBefore = int.Parse(lengthBeforeStr);
            if (markIx == str.Length - 1)
                return new ChangeSet { LengthBefore = lengthBefore, LengthAfter = 0 };
            var parts = str.Substring(markIx + 1).Split(',');
            var res = new ChangeSet { LengthBefore = lengthBefore, LengthAfter = parts.Length };
            foreach (string s in parts)
            {
                int val;
                if (int.TryParse(s, out val)) res.Items.Add(val);
                else res.Items.Add(new XieChar(s));
            }
            return res;
        }

        public string ToDiagStr()
        {
            var res = LengthBefore.ToString();
            res += '>';
            for (int i = 0; i < Items.Count; ++i)
            {
                if (i != 0) res += ',';
                if (Items[i] is XieChar)
                {
                    var hanzi = (Items[i] as XieChar).Hanzi;
                    res += hanzi == "\n" ? "\\n" : hanzi;
                }
                else res += ((int)Items[i]).ToString();
            }
            return res;
        }

        public bool IsValid()
        {
            if (LengthAfter != Items.Count) return false;
            int last = -1;
            foreach (var o in Items)
            {
                if (o is XieChar) continue;
                int val = (int)o;
                if (val < 0) return false;
                if (val > LengthBefore - 1) return false;
                if (val <= last) return false;
                last = val;
            }
            return true;
        }
        
        public static XieChar[] Apply(XieChar[] text, ChangeSet cs)
        {
            if (text.Length != cs.LengthBefore)
                throw new ArgumentException("Change set's lengthBefore must match text length");
            List<XieChar> items = new List<XieChar>(cs.LengthAfter);
            for (int ix = 0; ix < cs.Items.Count; ++ix)
            {
                if (cs.Items[ix] is XieChar) items.Add(cs.Items[ix] as XieChar);
                else items.Add(text[(int)cs.Items[ix]]);
            }
            return items.ToArray();
        }

        public static void ForwardPositions(ChangeSet cs, int[] poss)
        {
            var pp = new List<int>(poss.Length);
            foreach (int x in poss) pp.Add(x);
            int length = 0;
            for (int i = 0; i < cs.Items.Count; ++i)
            {
                if (cs.Items[i] is XieChar) ++length;
                else
                {
                    int ix = (int)cs.Items[i];
                    for (int j = 0; j < pp.Count; ++j)
                    {
                        if (pp[j] == -1) continue;
                        if (ix + 1 == pp[j])
                        {
                            poss[j] = length + 1;
                            pp[j] = -1;
                        }
                        else if (ix >= pp[j])
                        {
                            poss[j] = length;
                            pp[j] = -1;
                        }
                    }
                    ++length;
                }
            }
            for (int j = 0; j < pp.Count; ++j)
                if (pp[j] != -1)
                    poss[j] = length;
        }

        public static ChangeSet Compose(ChangeSet a, ChangeSet b)
        {
            if (a.LengthAfter != b.LengthBefore)
                throw new ArgumentException("LengthAfter of LHS must equal LengthBefore of RHS.");
            var res = new ChangeSet { LengthBefore = a.LengthBefore, LengthAfter = b.LengthAfter };
            res.Items.Capacity = res.LengthAfter;
            for (int i = 0; i < b.LengthAfter; ++i)
            {
                if (b.Items[i] is XieChar) res.Items.Add(b.Items[i]);
                else
                {
                    int ix = (int)b.Items[i];
                    res.Items.Add(a.Items[ix]);
                }
            }
            return res;
        }

        public static ChangeSet Merge(ChangeSet a, ChangeSet b)
        {
            if (a.LengthBefore != b.LengthBefore)
                throw new ArgumentException("Two change sets must have same LengthBefore.");
            ChangeSet res = new ChangeSet { LengthBefore = a.LengthBefore };
            int ixa = 0, ixb = 0;
            while (ixa < a.Items.Count || ixb < b.Items.Count)
            {
                if (ixa == a.Items.Count)
                {
                    if (b.Items[ixb] is XieChar) res.Items.Add(b.Items[ixb]);
                    ++ixb;
                    continue;
                }
                if (ixb == b.Items.Count)
                {
                    if (a.Items[ixa] is XieChar) res.Items.Add(a.Items[ixa]);
                    ++ixa;
                    continue;
                }
                // We got stuff in both
                XieChar ca = a.Items[ixa] as XieChar;
                XieChar cb = b.Items[ixb] as XieChar;
                // Both are kept characters: sync up position, and keep what's kept in both
                if (ca == null && cb == null)
                {
                    int vala = (int)a.Items[ixa];
                    int valb = (int)b.Items[ixb];
                    if (vala == valb)
                    {
                        res.Items.Add(vala);
                        ++ixa; ++ixb;
                        continue;
                    }
                    else if (vala < valb) ++ixa;
                    else ++ixb;
                    continue;
                }
                // Both are insertions: insert both, in lexicographical order (so merge is commutative)
                if (ca != null && cb != null)
                {
                    if (ca.CompareTo(cb) < 0)
                    {
                        res.Items.Add(ca);
                        res.Items.Add(cb);
                    }
                    else
                    {
                        res.Items.Add(cb);
                        res.Items.Add(ca);
                    }
                    ++ixa; ++ixb;
                    continue;
                }
                // If only one is an insertion, keep that, and advance in that changeset
                if (ca != null)
                {
                    res.Items.Add(ca);
                    ++ixa;
                }
                else
                {
                    res.Items.Add(cb);
                    ++ixb;
                }
            }
            res.LengthAfter = res.Items.Count;
            return res;
        }

        public static ChangeSet Follow(ChangeSet a, ChangeSet b)
        {
            if (a.LengthBefore != b.LengthBefore)
                throw new ArgumentException("Two change sets must have same LengthBefore.");
            ChangeSet res = new ChangeSet { LengthBefore = a.LengthAfter };
            int ixa = 0, ixb = 0;
            //int ofs = 0;
            while (ixa < a.Items.Count || ixb < b.Items.Count)
            {
                if (ixa == a.Items.Count)
                {
                    // Insertions in B become insertions
                    if (b.Items[ixb] is XieChar) res.Items.Add(b.Items[ixb]);
                    ++ixb;
                    continue;
                }
                if (ixb == b.Items.Count)
                {
                    // Insertions in A become retained characters
                    //if (a.Items[ixa] is XieChar) res.Items.Add(ixa + ofs);
                    if (a.Items[ixa] is XieChar) res.Items.Add(ixa);
                    ++ixa;
                    continue;
                }
                // We got stuff in both
                XieChar ca = a.Items[ixa] as XieChar;
                XieChar cb = b.Items[ixb] as XieChar;
                // Both are kept characters: sync up position, and keep what's kept in both
                if (ca == null && cb == null)
                {
                    int vala = (int)a.Items[ixa];
                    int valb = (int)b.Items[ixb];
                    if (vala == valb)
                    {
                        //res.Items.Add(vala + ofs);
                        res.Items.Add(ixa);
                        ++ixa; ++ixb;
                        continue;
                    }
                    else if (vala < valb) ++ixa;
                    else ++ixb;
                    continue;
                }
                // Insertions in A become retained characters
                // We consume A first
                if (ca != null)
                {
                    res.Items.Add(ixa);
                    ++ixa;
                    continue;
                }
                // Insertions in B become insertions
                else
                {
                    res.Items.Add(b.Items[ixb]);
                    ++ixb;
                    continue;
                }
            }
            res.LengthAfter = res.Items.Count;
            return res;
        }
    }
}

File: XiePinyin/client-source/editor/changeset.js
"use strict";

module.exports = (function () {

  function makeEmpty() {
    return {
      lengthBefore: 0,
      lengthAfter: 0,
      items: [],
    };
  }

  function makeIdent(length) {
    let res = {
      lengthBefore: length,
      lengthAfter: length,
      items: [],
    }
    for (let i = 0; i < length; ++i)
      res.items.push(i);
    return res;
  }

  function chrCmp(a, b) {
    if (a.hanzi.localeCompare(b.hanzi) != 0) return a.hanzi.localeCompare(b.hanzi);
    if (a.pinyin == b.pinyin) return 0;
    if (a.pinyin && !b.pinyin) return -1;
    if (!a.pinyin && b.pinyin) return 1;
    return a.pinyin.localeCompare(b.pinyin);
  }

  function isValid(cs) {
    if (cs.lengthAfter != cs.items.length) return false;
    let last = -1;
    for (let i = 0; i < cs.items.length; ++i)
    {
      const itm = cs.items[i];
      if (typeof itm === "object") continue;
      if (itm < 0) return false;
      if (itm > cs.lengthBefore - 1) return false;
      if (itm <= last) return false;
      last = itm;
    }
    return true;
  }

  function makeDiag(str) {
    let lengthBefore = str.substr(0, str.indexOf(">"));
    let res = {
      lengthBefore: Number(lengthBefore),
      items: [],
    };
    let rest = str.substr(str.indexOf(">") + 1);
    if (rest.length > 0) {
      let parts = rest.split(",");
      for (let i = 0; i < parts.length; ++i) {
        let part = parts[i];
        let val = parseInt(part, 10);
        if (isNaN(val)) res.items.push({ hanzi: part });
        else res.items.push(val);
      }
    }
    res.lengthAfter = res.items.length;
    return res;
  }

  function writeDiag(cs) {
    let res = "";
    res += cs.lengthBefore + ">";
    for (let i = 0; i < cs.items.length; ++i) {
      if (i > 0) res += ",";
      if (typeof cs.items[i] === "object") res += cs.items[i].hanzi;
      else res += cs.items[i];
    }
    return res;
  }

  function apply(text, cs) {
    if (text.length != cs.lengthBefore)
      throw "Change set's lengthBefore must match text length";
    let items = [];
    for (let ix = 0; ix < cs.items.length; ++ix) {
      if (typeof cs.items[ix] === "object") items.push(cs.items[ix]);
      else items.push(text[cs.items[ix]]);
    }
    return items;
  }

  function addReplace(cs, start, end, newText) {
    if (start < 0 || end < start)
      throw "bad values; expected: start >= 0 and end >= start";
    if (start > cs.lengthAfter || end > cs.lengthAfter)
      throw "start or end beyond lengthAfter of changeset";
    let csRepl = {
      lengthBefore: cs.lengthAfter,
      items: [],
    };
    for (let ix = 0; ix < start; ++ix) csRepl.items.push(ix);
    for (let i = 0; i < newText.length; ++i) csRepl.items.push(newText[i]);
    for (let ix = end; ix < cs.lengthAfter; ++ix) csRepl.items.push(ix);
    csRepl.lengthAfter = csRepl.items.length;
    return compose(cs, csRepl);
  }

  function compose(a, b) {
    if (a.lengthAfter != b.lengthBefore)
      throw "lengthAfter of LHS must equal lengthBefore of RHS";
    let res = {
      lengthBefore: a.lengthBefore,
      lengthAfter: b.lengthAfter,
      items: [],
    };
    for (let i = 0; i < b.lengthAfter; ++i)
    {
      if (typeof b.items[i] === "object") res.items.push(b.items[i]);
      else {
        let ix = Number(b.items[i]);
        res.items.push(a.items[ix]);
      }
    }
    return res;
  }

  function merge(a, b) {
    if (a.lengthBefore != b.lengthBefore)
      throw "The two change sets must have same lengthBefore";
    let res = {
      lengthBefore: a.lengthBefore,
      items: [],
    };
    let ixa = 0, ixb = 0;
    while (ixa < a.items.length || ixb < b.items.length) {
      if (ixa == a.items.length) {
        if (typeof b.items[ixb] === "object") res.items.push(b.items[ixb]);
        ++ixb;
        continue;
      }
      if (ixb == b.items.length) {
        if (typeof a.items[ixa] === "object") res.items.push(a.items[ixa]);
        ++ixa;
        continue;
      }
      // We got stuff in both
      let ca = null;
      if (typeof a.items[ixa] === "object") ca = a.items[ixa];
      let cb = null;
      if (typeof b.items[ixb] === "object") cb = b.items[ixb];
      // Both are kept characters: sync up position, and keep what's kept in both
      if (ca == null && cb == null) {
        let vala = a.items[ixa];
        let valb = b.items[ixb];
        if (vala == valb) {
          res.items.push(vala);
          ++ixa; ++ixb;
          continue;
        }
        else if (vala < valb) ++ixa;
        else ++ixb;
        continue;
      }
      // Both are insertions: insert both, in lexicographical order (so merge is commutative)
      if (ca != null && cb != null) {
        if (chrCmp(ca, cb) < 0) {
          res.items.push(ca);
          res.items.push(cb);
        }
        else {
          res.items.push(cb);
          res.items.push(ca);
        }
        ++ixa; ++ixb;
        continue;
      }
      // If only one is an insertion, keep that, and advance in that changeset
      if (ca != null) {
        res.items.push(ca);
        ++ixa;
      }
      else {
        res.items.push(cb);
        ++ixb;
      }
    }
    res.lengthAfter = res.items.length;
    return res;
  }

  function follow(a, b) {
    if (a.lengthBefore != b.lengthBefore)
      throw "The two change sets must have same lengthBefore";
    let res = {
      lengthBefore: a.lengthAfter,
      items: [],
    };
    let ixa = 0, ixb = 0;
    while (ixa < a.items.length || ixb < b.items.length) {
      if (ixa == a.items.length) {
        // Insertions in B become insertions
        if (typeof b.items[ixb] === "object") res.items.push(b.items[ixb]);
        ++ixb;
        continue;
      }
      if (ixb == b.items.length) {
        // Insertions in A become retained characters
        if (typeof a.items[ixa] === "object") res.items.push(ixa);
        ++ixa;
        continue;
      }
      // We got stuff in both
      let ca = null;
      if (typeof a.items[ixa] === "object") ca = a.items[ixa];
      let cb = null;
      if (typeof b.items[ixb] === "object") cb = b.items[ixb];
      // Both are kept characters: sync up position, and keep what's kept in both
      if (ca == null && cb == null) {
        let vala = a.items[ixa];
        let valb = b.items[ixb];
        if (vala == valb) {
          res.items.push(vala);
          ++ixa; ++ixb;
          continue;
        }
        else if (vala < valb) ++ixa;
        else ++ixb;
        continue;
      }
      // Insertions in A become retained characters
      // We consume A first
      if (ca != null) {
        res.items.push(ixa);
        ++ixa;
        continue;
      }
      // Insertions in B become insertions
      else {
        res.items.push(b.items[ixb]);
        ++ixb;
        continue;
      }
    }
    res.lengthAfter = res.items.length;
    return res;
  }

  return {
    makeEmpty,
    makeIdent,
    makeDiag,
    writeDiag,
    addReplace,
    apply,
    chrCmp,
    isValid,
    compose,
    merge,
    follow,
  };
})();
